src/workshed.ts

```diff
@@ -0,0 +1,390 @@
+
+export enum TrainsetPiece {
+  UNKNOWN = "",
+  EMPTY = "empty",
+  GAIN_MEAT = "meat_mine",
+  EFFECT_MP = "tower_fizzy",
+  GAIN_STATS = "viewing_platform",
+  HOT_RES_COLD_DMG = "tower_frozen",
+  STENCH_RES_SPOOKY_DMG = "spooky_graveyard",
+  SMUT_BRIDGE_OR_STATS = "logging_mill",
+  CANDY = "candy_factory",
+  DOUBLE_NEXT_STATION = "coal_hopper",
+  COLD_RES_STENCH_DMG = "tower_sewage",
+  SPOOKY_RES_SLEAZE_DMG = "oil_refinery",
+  SLEAZE_RES_HOT_DMG = "oil_bridge",
+  MORE_ML = "water_bridge",
+  MOXIE_STATS = "groin_silo",
+  RANDOM_BOOZE = "grain_silo",
+  MYS_STATS = "brain_silo",
+  MUS_STATS = "brawn_silo",
+  BUFF_FOOD_DROP = "prawn_silo",
+  DROP_LAST_FOOD_OR_RANDOM = "trackside_diner",
+  ORE = "ore_hopper",
+}
+
+const pieces: TrainsetPiece[] = [
+  TrainsetPiece.EMPTY,
+  TrainsetPiece.GAIN_MEAT,
+  TrainsetPiece.EFFECT_MP,
+  TrainsetPiece.GAIN_STATS,
+  TrainsetPiece.HOT_RES_COLD_DMG,
+  TrainsetPiece.STENCH_RES_SPOOKY_DMG,
+  TrainsetPiece.SMUT_BRIDGE_OR_STATS,
+  TrainsetPiece.CANDY,
+  TrainsetPiece.DOUBLE_NEXT_STATION,
+  TrainsetPiece.COLD_RES_STENCH_DMG,
+  TrainsetPiece.UNKNOWN,
+  TrainsetPiece.SPOOKY_RES_SLEAZE_DMG,
+  TrainsetPiece.SLEAZE_RES_HOT_DMG,
+  TrainsetPiece.MORE_ML,
+  TrainsetPiece.MOXIE_STATS,
+  TrainsetPiece.RANDOM_BOOZE,
+  TrainsetPiece.MYS_STATS,
+  TrainsetPiece.MUS_STATS,
+  TrainsetPiece.BUFF_FOOD_DROP,
+  TrainsetPiece.DROP_LAST_FOOD_OR_RANDOM,
+  TrainsetPiece.ORE,
+];
+
+function getPieceId(piece: TrainsetPiece): number {
+  return Math.max(0, pieces.indexOf(piece));
+}
+
+function getTrainsetPositionsUntilConfigurable(): number {
+  const pos = toInt(get("trainsetPosition", ""));
+  const configured = toInt(get("lastTrainsetConfiguration", ""));
+  const turnsSinceConfigured = pos - configured;
+
+  return Math.max(0, 40 - turnsSinceConfigured);
+}
+
+function isTrainsetConfigurable(): boolean {
+  // eslint-disable-next-line libram/verify-constants
+  return getWorkshed() === $item`model train set` && getTrainsetPositionsUntilConfigurable() <= 0;
+}
+
+function setTrainsetConfiguration(pieces: TrainsetPiece[]) {
+  visitUrl("campground.php?action=workshed");
+
+  const pieceIds = pieces
+    .map((p) => getPieceId(p))
+    .map((pieceId, index) => `slot[${index}]=${pieceId}`);
+
+  const url = `choice.php?forceoption=0&whichchoice=1485&option=1&pwd&${pieceIds.join("&")}`;
+
+  visitUrl(url, true);
+  visitUrl("main.php");
+
+  const expected = pieces.join(",");
+
+  if (expected !== get("trainsetConfiguration", "")) {
+    throw new Error(
+      `Expected trainset configuration to have changed, expected "${expected}" but instead got ${get(
+        "trainsetConfiguration",
+        ""
+      )}`
+    );
+  }
+}
+
+class GarboWorkshed {
+  workshed: Item;
+  done: () => boolean;
+  action: () => void;
+
+  constructor(workshed: Item, done: () => boolean, action: () => void) {
+    this.workshed = workshed;
+    this.done = done;
+    this.action = action;
+  }
+}
```

Comment:
Right now I don't see any particular reason this should be a Class rather than just a typed object, given that it has no methods

Reply:
Cleaned things up by using default functions, which interfaces do not support